### PR TITLE
fix: allow rename of structuremetadata labels

### LINKED ISSUE #13937: Label filters ignore parsed label if it exists in structured metadata
**Describe the bug**
When a line has a label in structured metadata, and the label is then overwritten by a label format expression, the label filter uses the structured metadata instead of the updated value.

**To Reproduce**
1. Started Loki (SHA or version): v3.1.1
2. Sent record to `/loki/api/v1/push`:

```
{
  "streams": [
    {
      "stream": {"service_name":"test"},
      "values": [["1724169776000000000", "hello", {"x": "foo"}]]
    }
  ]
}
```

Command:
```
curl -H 'Content-Type: application/json' -d '{"streams": [{"stream": {"service_name":"test"},"values":[["1724169776000000000", "hello", {"x": "foo"}]]}]}' -v 'http://127.0.0.1:3100/loki/api/v1/push'
```

4. Sent queries:

- `{service_name="test"} | label_format x="bar" | x="foo"`
- `{service_name="test"} | label_format x="bar" | x="bar"`

Commands:

```
curl -s -G http://127.0.0.1:3100/loki/api/v1/query_range --data-urlencode start=1724169776000000000 --data-urlencode query='{service_name="test"} | label_format x="bar" | x="foo"' | jq '.data.result[]'
curl -s -G http://127.0.0.1:3100/loki/api/v1/query_range --data-urlencode start=1724169776000000000 --data-urlencode query='{service_name="test"} | label_format x="bar" | x="bar"' | jq '.data.result[]'
```

5. Observe that the first query (the one with the `x="foo"` filter) produces a result with `x="bar"`, and the second query (the one with the `x="bar"` filter) produces no result.

**Expected behavior**
Given that `label_format` is supposed to always set `x` to `"bar"`, only the second query (`... | label_format x="bar" | x="bar"`) should ever produce a result. The first query (`... | label_format x="bar" | x="foo"`) should never produce anything, as `"bar" != "foo"`.

**Environment:**
 - Infrastructure: bare-metal
 - Deployment tool: `/bin/bash`

**Screenshots, Promtail config, or terminal output**
```
localhost ~ # curl -H 'Content-Type: application/json' -d '{"streams": [{"stream": {"service_name":"test"},"values":[["1724169776000000000", "hello", {"x": "foo"}]]}]}' 'http://127.0.0.1:3100/loki/api/v1/push'
localhost ~ # curl -s -G http://127.0.0.1:3100/loki/api/v1/query_range --data-urlencode start=1724169776000000000 --data-urlencode query='{service_name="test"} | label_format x="bar" | x="foo"' | jq '.data.result'
[
  {
    "stream": {
      "service_name": "test",
      "x": "bar"
    },
    "values": [
      [
        "1724169776000000000",
        "hello"
      ]
    ]
  }
]
localhost ~ # curl -s -G http://127.0.0.1:3100/loki/api/v1/query_range --data-urlencode start=1724169776000000000 --data-urlencode query='{service_name="test"} | label_format x="bar" | x="bar"' | jq '.data.result'
[]
```

### PR DESCRIPTION
**Pull request by ravishankar15**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
Fixes #13937

**Special notes for your reviewer**:

**Checklist**
- [x] Reviewed the [`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md) guide (**required**)
- [x] Documentation added
- [x] Tests updated
- [x] Title matches the required conventional commits format, see [here](https://www.conventionalcommits.org/en/v1.0.0/)
  - **Note** that Promtail is considered to be feature complete, and future development for logs collection will be in [Grafana Alloy](https://github.com/grafana/alloy). As such, `feat` PRs are unlikely to be accepted unless a case can be made for the feature actually being a bug fix to existing behavior.
- [x] Changes that require user attention or interaction to upgrade are documented in `docs/sources/setup/upgrade/_index.md`
- [x] For Helm chart changes bump the Helm chart version in `production/helm/loki/Chart.yaml` and update `production/helm/loki/CHANGELOG.md` and `production/helm/loki/README.md`. [Example PR](https://github.com/grafana/loki/commit/d10549e3ece02120974929894ee333d07755d213)
- [x] If the change is deprecating or removing a configuration option, update the `deprecated-config.yaml` and `deleted-config.yaml` files respectively in the `tools/deprecated-config-checker` directory. [Example PR](https://github.com/grafana/loki/pull/10840/commits/0d4416a4b03739583349934b96f272fb4f685d15)

### CODE REVIEW

**Review by chaudum — COMMENTED** [review]
Thanks for your contribution!

File: pkg/logql/log/fmt.go
Comment (salvacorts) [reply pkg/logql/log/fmt.go:410]: Thank you @ravishankar15 for creating the PR!

> @salvacorts Maybe I am overly cautious and my worries are ungrounded, so I want your opinion on this.

I agree @chaudum, this problem may not only limited to the label formatter, but to any pipeline expression that may overwrite a label that already exists in structured metadata.

The bug is in the [`LabelsBuilder.UnsortedLabels`](https://github.com/grafana/loki/blob/4c88be0ef2310406206c02d7747bf73b51c95bda/pkg/logql/log/labels.go#L406) method, which seems to be used only by `IntoMap` (later used by `LabelsFormatter.Process`). This is what I'd do:

- Similarly to [how we skip stream labels that will be replaced][1], skip structured metadata fields that will be replaced.
- The [`BaseLabelsBuilder.add` field][2] is an array of 3 items (one per category). Each item is a list of labels to add for the category. I would replace this by two new fields:
    - `addStructuredMetadata` and `addParsed`
    - Then in  LabelsBuilder.UnsortedLabels, we can filter out stream labels looking at `addStructuredMetadata` and `addParsed`, and then filter out structured metadata looking at the `addParsed`

On top of this, I'd add tests to the [labels test file](pkg/logql/log/labels_test.go)

How confortable would you feel making these changes @ravishankar15? We are happy to help!

[2]: https://github.com/grafana/loki/blob/4c88be0ef2310406206c02d7747bf73b51c95bda/pkg/logql/log/labels.go#L130
[1]: https://github.com/grafana/loki/blob/4c88be0ef2310406206c02d7747bf73b51c95bda/pkg/logql/log/labels.go#L433-L452

Comment (ravishankar15) [reply pkg/logql/log/fmt.go:410]: Thanks for taking a look at the PR will check these out over the weekend

Comment (ravishankar15) [reply pkg/logql/log/fmt.go:410]: As @chaudum  mentioned I think there are two parts of the problem.

1. The suggestion from Salvacorts will solve the problem where we will not have duplicate values in the response
2. For `label filter after a label format expression needs to match the newly assigned value.` as mentioned by Chandum we need to make changes to the `getWithCategory` method on that note we are setting the flag `referencedStructuredMetadata` to true if the label is from `StructuredMetadata` in our issue the label is from `StructuredMetadata` but its renamed by LabelFormat in `ParsedLabel` so if we make the change the flag will be false I hope that should be okay and expected behaviour.
3. Also I believe we should handle the case in `Add` and `Set` method of LabelsBuilder ? 

I believe we need to solve both the issues for this bug to be resolved truly. Can you clarify my (2) and (3) and I can change my PR accordingly. 

Also, In @salvacorts suggestion was to move away from having single field `BaseLabelsBuilder.add` implementation and have separate fields `addStructuredMetadata` and `addParsed`. Do you feel we need to combine both the changes in a single PR. I feel it will be little cumbersome If its okay shall we split it into two PR's like

1. Fixes the two issue mentioned in chandum comment -> With existing `BaseLabelsBuilder.add`  implementation
2. Modify the `BaseLabelsBuilder.add` implementation to use separate fields `addStructuredMetadata` and `addParsed`

Comment (ravishankar15) [reply pkg/logql/log/fmt.go]: I have made the changes to the builder as per our discussion without reformatting which I am hoping should fix the issue.

Also @salvacorts I see the StreamLabels is not used logically outside of builder but I see some tests using that category to add labels and test some scenarios are all those outdated ?  Let me know I can raise a separate PR as a follow up to remove all those stale logic and clean up the builder.

**Comment by ravishankar15** [comment]
I have made the changes to the builder as per our discussion without reformatting which I am hoping should fix the issue.

Also @salvacorts I see the StreamLabels is not used logically outside of builder but I see some tests using that category to add labels and test some scenarios are all those outdated ?  Let me know I can raise a separate PR as a follow up to remove all those stale logic and clean up the builder.

**Comment by ravishankar15** [comment]
Hi Team, Is something else required in this PR ?

**Review by chaudum — APPROVED** [review]
LGTM
@salvacorts Can I have a second pair of eyes on this PR?

File: pkg/logql/log/labels_test.go
Comment (salvacorts) [inline pkg/logql/log/labels_test.go:364]: nit: you can use `require.ElementsMatch()` and skip the sorts

**Review by salvacorts — APPROVED** [review]
Looks great 👏 I left only a comment to add some comments to the `Set` method, and a nit for the test (feel free to disregard it).

I'm approving to unblock. Thank you tons for going the extra mile here :)

File: pkg/logql/log/labels.go
Comment (ravishankar15) [reply pkg/logql/log/labels.go:488]: @salvacorts Once this PR is merged I am planning to raise a separate PR,  Removing the stream label and separating the `add` object as discussed here https://github.com/grafana/loki/pull/13955#discussion_r1735713847

**Review by salvacorts — APPROVED** [review]

**Comment by salvacorts** [comment]
PR merged, thank you @ravishankar15 👏

**Pull request merged**
